### PR TITLE
Add check for zero vector in angle calculation of DetectorInfo

### LINKED ISSUE #29103: Find long term solution to dummy detectors in polaris
In the workspaces of polaris have some spectra assigned to dummy detectors. these detectors have no geometry and cause problems in unit convertion since they are not ignored lime monitors are. whenever this has come up before I have manualy had a work around but there should be a safer way of handeling this.

### PR DESCRIPTION
**Pull request by Harrietbrown**
**Description of work.**

This Commit adds a check in detector info that catching runtime errors from V3D when calculation angles. This error appears often when working with polaris data which features detectors with l1 and l2 equal to zero but are not monitors.

**Report to:** Helen Playford

**To test:**

The script below loads a raw polaris workspace, the instrument has dummy detectors without geometry but not marked as monitors. The `ConvertUnits` algorithm should give warnings that angles can not be calculated, but it should still run. 
```
from mantid.simpleapi import *

Load(Filename=r'\\isis.cclrc.ac.uk\inst$\ndxpolaris\instrument\data\cycle_17_3\POLARIS00100017.nxs', OutputWorkspace='POLARIS00100017')
ConvertUnits(InputWorkspace='silicon_raw', OutputWorkspace='merged_ws', Target='MomentumTransfer')
```

Fixes #29103 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by alicerussell1 — CHANGES_REQUESTED** [review]
This works well, would you be able to add a unit test for this?

**Review by alicerussell1 — DISMISSED** [review]
Looks good to me 👍

File: Framework/Geometry/src/Instrument/DetectorInfo.cpp
Comment (martyngigg) [inline Framework/Geometry/src/Instrument/DetectorInfo.cpp:162]: I'm going to have to hold this up for the moment I'm afraid as I think this is too low-level a place to do this. As `0.0` could be a real angle then from the outside it's impossible for a user of this method to tell if `0.0` means it really is 0.0 or an error occurred. As this is low level then some other user of the method may want to do to do something different if the exceptional case happens.

Can I ask if we know why the POLARIS IDF has these detectors and is it just `ConvertUnits` that fails?

**Review by martyngigg — CHANGES_REQUESTED** [review]

File: Framework/Geometry/src/Instrument/DetectorInfo.cpp
Comment (martyngigg) [inline Framework/Geometry/src/Instrument/DetectorInfo.cpp:164]: I'm still a bit uncomfortable with us catching this at such a low level. The other parts of this function still throw if there are errors. A user of this function would now have to check for both an exception and a `NaN` to confirm there was no error and it would pass silently.

Is the error that Polaris are seeing just with`ConvertUnits`? Can we just catch the errors there and handle them? Other users of this code may want to handle this situation differently and would then just have to catch any exceptions that indicate an error and handle it accordingly.

Comment (Harrietbrown) [reply Framework/Geometry/src/Instrument/DetectorInfo.cpp]: I've moved the check into convert units. If we need to get twotheta from polaris we will need to have these checks.

File: Framework/Algorithms/src/ConvertUnits.cpp
Comment (martyngigg) [inline Framework/Algorithms/src/ConvertUnits.cpp:421]: Thanks for changing this. This way it has no effect on performance when `twoTheta` doesn't throw.

Just one minor cleanup: Creating a `Logger` is quite expensive and the Algorithm base class already creates one for each algorithm that is run so you should just be able to remove `Kernel::Logger g_log("ConvertUnits");` and `g_log.warning` will use the one from the base class.

File: Framework/Algorithms/test/ConvertUnitsTest.h
Comment (martyngigg) [inline Framework/Algorithms/test/ConvertUnitsTest.h:677]: Good work on adding a new test for this.

**Review by martyngigg — CHANGES_REQUESTED** [review]

**Review by martyngigg — APPROVED** [review]
This looks good now. Good to go when Jenkins gives the go ahead.

**Pull request merged**
